### ISSUE TITLE
Failed to Login > Error setting up platform netgear (More Details Enclosed)

### ISSUE THREAD

**Issue by caffeinatedMike** [MatMaul/pynetgear#47]
**Platform**
Old Laptop running fresh install of Linux Mint, running Home Assistant in Docker. Installed HA using [these directions](https://gist.github.com/frenck/32b4f74919ca6b95b30c66f85976ec58), with a few minor detours
 1. Needing to update python3, since Linux Mint natively comes with 3.5.3. Followed [this guide](https://mintguide.org/other/794-python-3-6-install-latest-version-into-linux-mint.html) to install python 3.6 along side the typical python3 installation of mint as to not disturb any of linux mint's other functionality.
 2. Needing to `pip install docker.io`, the default docker library for linux mint, instead of the docker mentioned in the gist
 3. Running each command from the gist individually instead of running the script as a whole.

**Home Assistant Version**
0.79.3

**Router Model**
Netgear [C3000](https://www.netgear.com/home/products/networking/cable-modems-routers/C3000.aspx)

**Problem**
Simply put, cannot get Netgear component to work as device_tracker. Fails to login. I've confirmed the credentials are correct. No matter what I try I cannot seem to get the component configured.

**Setup Variations I've tried in Configuration.yaml**
```
device_tracker:
  - platform: netgear
    host: 192.168.0.1
    port: 80
    username: admin
    password: REDACTED
```
```
device_tracker:
  - platform: netgear
    host: 192.168.0.1
    username: admin
    password: REDACTED
```
```
device_tracker:
  - platform: netgear
    password: REDACTED
```

**Full Log**
```
2018-10-04 08:04:46 ERROR (MainThread) [homeassistant.components.device_tracker] Unable to load /config/known_devices.yaml: Config file not found: /config/known_devices.yaml
2018-10-04 08:04:47 ERROR (SyncWorker_3) [pynetgear] Error talking to API
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 384, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 380, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/local/lib/python3.6/http/client.py", line 1331, in getresponse
    response.begin()
  File "/usr/local/lib/python3.6/http/client.py", line 297, in begin
    version, status, reason = self._read_status()
  File "/usr/local/lib/python3.6/http/client.py", line 258, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/local/lib/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
ConnectionResetError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 445, in send
    timeout=timeout
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 638, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/local/lib/python3.6/site-packages/urllib3/util/retry.py", line 367, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/local/lib/python3.6/site-packages/urllib3/packages/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 384, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "/usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 380, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/local/lib/python3.6/http/client.py", line 1331, in getresponse
    response.begin()
  File "/usr/local/lib/python3.6/http/client.py", line 297, in begin
    version, status, reason = self._read_status()
  File "/usr/local/lib/python3.6/http/client.py", line 258, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/local/lib/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/pynetgear/__init__.py", line 247, in _make_request
    data=message, timeout=30, verify=False)
  File "/usr/local/lib/python3.6/site-packages/requests/api.py", line 112, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 512, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 622, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/adapters.py", line 495, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
2018-10-04 08:04:47 ERROR (SyncWorker_3) [homeassistant.components.device_tracker.netgear] Failed to Login
2018-10-04 08:04:47 ERROR (MainThread) [homeassistant.components.device_tracker] Error setting up platform netgear
018-10-04 12:15:55 WARNING (SyncWorker_0) [netdisco.ssdp] Found malformed XML at http://192.168.0.10:9080: status=ok
```

**Edit**
I should also note that the laptop running the HA instance is connected directly to the router via Ethernet cable.
**Edit 2**
I notice my router isn't included on the [Supported Routers List](https://github.com/MatMaul/pynetgear#supported-routers), but I can access & manage the router using the Netgear Genie android app and my device is listed in the acceptable devices in the nodejs version of this program that this was ported over from.

CC: @MatMaul

**Comment by caffeinatedMike** [comment]
I found a Node.js implementation that allows access to my router [here](https://github.com/gruijter/netgear.js/), but it doesn't help since I need to have Home Assistant recognize it, so I'm hoping I can get some help with pynetgear to successfully incorporate it.

**Comment by MatMaul** [comment]
It seems like it can't connect at all to the soap interface.
Is your router web interface running on port 80 ?
You should play with ssl parameter (boolean) too. Which I forgot to document :( bad me.

**Comment by MatMaul** [comment]
I don't think I can help you with that, pretty sure it's on your side. It still completely fails to connect to the router, why I don't know.
Use your rPi first, you are on Windows, I am not even going further with that :)

**Comment by caffeinatedMike** [comment]
Haha, I don't understand why it would automatically detect the Gateway part of the router, but not the Router? I mean, everything setting-wise on the router side looks fine.
![1](https://i.imgur.com/BpaNIN4.jpg)
![2](https://i.imgur.com/sNY5mou.jpg)
Do you have *any* idea what could be preventing it from working?

**Comment by caffeinatedMike** [comment]
I just got done doing a packet sniffing session on my Android device while (successfully) the Netgear Genie app to view attached devices. If you want I'd be willing to provide the entire packet session in an attempt to help you debug why this module isn't currently working with my router.

**Comment by caffeinatedMike** [comment]
@MatMaul I've emailed you 3 text files of http calls (login soap actions & list device actions). Hope it helps diagnose the issue.

**Comment by caffeinatedMike** [comment]
Just noting the one difference I was able to find when examining the PyNetgear's attempted communications with my router and what the Netgear Genie app shows. The only difference I found (and what I believe the be the issue, at least in my case) is as follows:

# PyNetgear's Login XML, used to gain access to Router's SOAP API
```
<SOAP-ENV:Body>
<Authenticate>
  <NewUsername>{username}</NewUsername>
  <NewPassword>{password}</NewPassword>
</Authenticate>
</SOAP-ENV:Body>
```
# My router's Login XML, captured with [Packet Capture](https://apkpure.com/packet-capture/app.greyshirts.sslcapture) android app.
```
<SOAP-ENV:Body>
<M1:SOAPLogin xmlns:M1="urn:NETGEAR-ROUTER:service:DeviceConfig:1">
  <Username>{username}</Username>
  <Password>{password}</Password>
</M1:SOAPLogin>
</SOAP-ENV:Body>
```

Since the login XML Schema appears to be different for my router that would explain why the component fails to communicate with my router at all.

**Comment by caffeinatedMike** [comment]
Will do. I'll modify a local copy of the module once I'm home from work and have direct access to the router. If that ends up being the culprit how would you suggest a change be implemented. Would you want me to add a try/except statement in the login function to support both login scenerios? I suspect that my router uses a legacy login flow (seeing as it's not as new as the Nighthawks or N7000s that people report having success with). Or do you want me to just report back if the testing was successful and leave it to you to implement to keep your logic consistent?

**Comment by MatMaul** [comment]
I would say try the current one first and if it fails try SOAPLogin. You should be able to do it without changing anything outside of login method.

**Comment by MatMaul** [comment]
And I am lazy so I'll gladly take a tested pull request.
I insist on tested ;)

**Comment by caffeinatedMike** [comment]
Haha, will do! Just wanted to make sure it wouldn't be considered too hacky of an implementation. But, if all you care about is if it's tested that's fine by me :)

**Comment by MatMaul** [comment]
@gruijter I think it makes sense (which is kind of unusual with this API :-P ), the SOAPAction is the full qualifier of the action, while in the xml you specify the context (DeviceConfig:1) and the action is the actual XML tag.

**Comment by MatMaul** [comment]
Nevermind you were saying the same ^^

**Comment by MatMaul** [comment]
@caffeinatedMike you can use SERVICE_DEVICE_CONFIG as service parameter when calling _make_request since it is already defined.

**Comment by gruijter** [comment]
behaviour seems to be different though. I just changed it to ^^ In my node package. Now I can only do one action, and ater that it seems to be logged out again. Need to check what is going on...

**Comment by MatMaul** [comment]
You mean that your router can take both login method but they behave differently ? My god this API is such a garbage...

**Comment by gruijter** [comment]
yes, my R7800 takes both methods. My genie app uses what @caffeinatedMike found though.

**Comment by MatMaul** [comment]
Ok then I'll try this one on my Orbi to see.
We can perhaps put the other one first if it works for a majority of models, and call it each time before doing the action call.

**Comment by caffeinatedMike** [comment]
Oh man, just saw all these comments over since my last reply. It's so good to hear that it's not just me being crazy. The api is just a f***** mess. So, if I'm understanding you correctly, both my and the current methods will be implemented (after your tests, if successful) to broaden the model-coverage of the component. Are you saying I _might_ actually finally be able to use my router as a device_tracker? :smiley:

**Comment by gruijter** [comment]
@MatMaul I found out that with this 'new' method for each request you must use the cookie  that you get on first login. But as always (very nice) Netgear has made the cookie header different on different models. For R7800 it is 'Set-Cookie', and on R7000 it is 'set-cookie'. It seems to work now in my node version. But I need to test some more. If it works I will probably deprecate the old login method, since this new method seems to be more consistent.

**Comment by gruijter** [comment]
here an example of login where you receive the cookie:

```
HTTP/1.1 200 OK
CONTENT-LENGTH: 426
Set-Cookie: sess_id=8696e1659b88558de96329c07b59ba95ad80a218ca1d9ea7c1605d8ad33032e85c4095f6afc959120959ff55e97857cdf163d64f0ffcdd4e5a311994959da540; SameSite=Strict
CONTENT-TYPE: text/xml; charset="UTF-8"
SERVER: "OS/version" UPnP/1.0 "product/version"

<?xml version="1.0" encoding="UTF-8"?>
<soap-env:Envelope
        xmlns:soap-env="http://schemas.xmlsoap.org/soap/envelope/"
        soap-env:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/"
        >
<soap-env:Body>
    <m:SOAPLoginResponse
        xmlns:m="urn:NETGEAR-ROUTER:service:DeviceConfig:1">
    </m:SOAPLoginResponse>
    <ResponseCode>000</ResponseCode>
</soap-env:Body>
</soap-env:Envelope>
```

**Comment by gruijter** [comment]
@caffeinatedMike If you want to try the new login method on node, you can use my beta branch here:
https://github.com/gruijter/netgear.js/tree/beta

**Comment by caffeinatedMike** [comment]
I don't have node, I'm only setup with python atm :/

**Comment by gruijter** [comment]
I made my version so that it first does getCurrentSettings (can be done without login). It contains a loginMethod which is 2.0 for both my R7000 and R7800. I assume this method defines if you can use the new login method, or that you have to use the old one (all assumptions now...)

Anyway, I made it to use the new login method only if the getCurrentSettings.loginMethod >= 2

Enough work for today :)
over and out

**Comment by caffeinatedMike** [comment]
Wait, how do I test any of this for myself (in my home assistant setup)? I know nothing about node lol

And side note: My login method also states 2.0 >.<

**Comment by gruijter** [comment]
Lol. I know nothing on HA or python😂

**Comment by caffeinatedMike** [comment]
Lol whelp that leaves me at a loss :laughing: I'll just try @MatMaul's suggestion
> You can change line 87/88 and LOGIN_BODY at line 380 to match your finding and see if it helps.
And see how that goes. Then, I'll report back and we can go from there :smile:

**Comment by MatMaul** [comment]
@gruijter can you show me an example of a request using the cookie ? I can't make it works while it looks good to me. Getting the cookie I managed to do but I can't use it.

**Comment by gruijter** [comment]
This is what I do for the header on each request. Only put in the cookie if you previously got one, or it will fail.

```const headers = {
				SOAPAction: action,
				'Cache-Control': 'no-cache',
				'User-Agent': 'node-netgearjs',
				'Content-Length': Buffer.byteLength(message),
				Connection: 'Keep-Alive',
			};
			if (this.cookie) {
				headers.Cookie = this.cookie;
			}
```

**Comment by caffeinatedMike** [comment]
@MatMaul To grab a the Set-Cookie header from the login response you'd have to add something after [Line 271](https://github.com/MatMaul/pynetgear/blob/master/pynetgear/__init__.py#L271) like so
`sess_cookie = req.headers['set-cookie']`

**Comment by MatMaul** [comment]
@gruijter Awesome thanks it's working now, I saw your code but I misinterpreted it.
I was setting Set-Cookie and not Cookie in the request headers.
I need to clean that up, we'll see tomorrow.

**Comment by caffeinatedMike** [comment]
@MatMaul I'll make a couple of adjustments to my fork for you to review. That way I can redeem myself a lot bit for making your life a living hell these last two days :stuck_out_tongue: 

I'll switch to using request.Session to make the cookie easier to manage & test out the previously discussed fixes.

**Comment by gruijter** [comment]
Here an example of logout command with use of the cookie
``` 

************ SOAPLogout **************
POST http://routerlogin.net/soap/server_sa/ HTTP/1.1
Accept: text/xml
SOAPAction: urn:NETGEAR-ROUTER:service:DeviceConfig:1#SOAPLogout
Cache-Control: no-cache
Pragma: no-cache
Accept-Encoding: gzip
Accept-Language: zh-CN,en,*
Content-type: multipart/form-data
User-Agent: SOAP Toolkit 3.0
Cookie: jwt_local=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJleHAiOiIxMDY2OTkiLCJpc3MiOiJ3d3cubmV0Z2Vhci5jb20iLCJzdWIiOiJTT0FQIFRvb2xraXQgMy4wIn0=.9d780d44a9d2d7fa95758fd9b3b4f37f1b8a60984ebb058f331344821a4e5163
Content-Length: 531
Host: routerlogin.net
Connection: Keep-Alive

<?xml version="1.0" encoding="utf-8" standalone="no"?>
<SOAP-ENV:Envelope xmlns:SOAPSDK1="http://www.w3.org/2001/XMLSchema" xmlns:SOAPSDK2="http://www.w3.org/2001/XMLSchema-instance" xmlns:SOAPSDK3="http://schemas.xmlsoap.org/soap/encoding/" xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/">
<SOAP-ENV:Header>
<SessionID>A7D88AE69687E58D9A00</SessionID>
</SOAP-ENV:Header>
<SOAP-ENV:Body>
<M1:SOAPLogout xmlns:M1="urn:NETGEAR-ROUTER:service:DeviceConfig:1">
</M1:SOAPLogout>
</SOAP-ENV:Body>
</SOAP-ENV:Envelope>

===========================
HTTP/1.0 200 OK
Content-Length: 385
Content-Type: text/xml; charset="UTF-8"
Server: Linux/2.6.15 uhttpd/1.0.0 soap/1.0

<?xml version="1.0" encoding="UTF-8"?>
<SOAP-ENV:Envelope
   xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/"
   SOAP-ENV:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
<SOAP-ENV:Body>
<m:SOAPLogoutResponse xmlns:m="urn:NETGEAR-ROUTER:service:DeviceConfig:1"></m:SOAPLogoutResponse>
<ResponseCode>000</ResponseCode>
</SOAP-ENV:Body>
</SOAP-ENV:Envelope>

```

**Comment by MatMaul** [comment]
I have the code working guys thanks. @caffeinatedMike if you want I push my branch and you can clean up my mess, and handle gracefully the 2 different ways to login.

**Comment by gruijter** [comment]
Super! Together we'll bring that netgear soap mess down on its knees :+1:

**Comment by MatMaul** [comment]
@caffeinatedMike here it is:
https://github.com/MatMaul/pynetgear/tree/new-login

It's clean but the fallback to the old login method is missing.

**Comment by caffeinatedMike** [comment]
Yay! I'll have a look as soon as I'm home (about half hour from now) & clean it up.

**Comment by MatMaul** [comment]
Ok give me 10mn adding some missing retry after failing that I removed accidentally.

**Comment by caffeinatedMike** [comment]
Ok, I'm home now. So, I'll start fiddling whenever you get done.

**Comment by MatMaul** [comment]
Done. The code is clean, you have the login_v1 and login_v2 methods, you just need to put some try code in login.

**Comment by caffeinatedMike** [comment]
Awesome! I'm on it.

**Comment by MatMaul** [comment]
@caffeinatedMike if you have some free time after your setup works you can try implementing and test new methods to mess with the router settings, like SetBlockDeviceByMAC. It should be a matter of copy-pasting and adapting some stuff from the existing ones.

There are some pointers here and you can have more details in @gruijter js code.
https://github.com/MatMaul/pynetgear/issues/20

**Comment by caffeinatedMike** [comment]
I'll see what I can do. Looking at your new-login branch I'm not seeing the login_v2 method?

**Comment by MatMaul** [comment]
Hint: so your tests are meaningful you can do v1 then v2, to test the fullback mechanism since your router fails with v1. Then once the retry logic is tested please put v2 as the first method to try. I can't easily test myself, both works here.

**Comment by caffeinatedMike** [comment]
I'm failing even with v2 atm :(

**Comment by caffeinatedMike** [comment]
Bahahahahahahahahahaha. @MatMaul you know what the issue's been all along? You never set the `Content-Type` header in the `_get_soap_headers` function :laughing: 

Added it, then :boom: Device List showed in IDLE/Shell
![Device List](https://i.imgur.com/BMOcNs7.jpg)

**Comment by caffeinatedMike** [comment]
Pull request submitted :smiley: 

To be honest, there's a lot that can be made more efficient in the module, but we can save that for another day :wink: 

About how long do you estimate these changes will take to make their way to the master?

**Comment by MatMaul** [comment]
@caffeinatedMike I'll probably release a new version this weekend. I am adding proper logging so we can get useful debug logs when debugging.

Can you test what fields are really necessary for you in the header ?
I am mainly interesting in removing `Connection: Keep-Alive` because I think we don't do keep-alive so we are kind of lying :)

If you can also test with `multipart/form-data` as a content type as mentioned by @gruijter .

**Comment by caffeinatedMike** [comment]
gruijter At second look, my captured genie calls also use multipart/form-data . So, I suspect the Content-Type will work with both. I'm thinking the header just needs to be set really. But, I'll switch the type later and test :)

MatMaul Of course, I'll test all that when I get home from work today and resubmit the PR tonight after work :)

**Comment by caffeinatedMike** [comment]
@MatMaul Here's the [updated commit](https://github.com/MatMaul/pynetgear/pull/49/commits/7ee97c39326842247c3f121bdf20e4378c2e03f5) to the PR in case you're able to test it within the next 8 hours. If you don't get around to it that's fine, I'll test when I get home in that case.

**Comment by caffeinatedMike** [comment]
@MatMaul I have tested the changes I made and I'm happy to say they are all functional. So, you can merge that PR whenever you have the time :smile: 

@gruijter I've been trying to run your test for you, but I keep failing. First, I'll point out one issue I was able to correct: you have an extra comma on line 22 `process.argv[5],`. Removing it got rid of the error. However, after that I receive a `SyntaxError: Unexpected token function`, which points to line 26 `async function getRouterInfo() {`

**Comment by MatMaul** [comment]
merged, and I just added my logging changes. I'll merge another PR and release a new version this weekend. Please continue testing with the new commits, always helpful to validate a bit before release, and the more the better :)

**Comment by caffeinatedMike** [comment]
I've been messing around with it in HA by copying the netgear component over into the `custom_components` directory, then adding my local pynetgear module inside there. Changed the import statements to point locally and been discovering/renaming my devices. So, I say so far so good :)

**Comment by gruijter** [comment]
> @MatMaul I have tested the changes I made and I'm happy to say they are all functional. So, you can merge that PR whenever you have the time 😄
> 
> @gruijter I've been trying to run your test for you, but I keep failing. First, I'll point out one issue I was able to correct: you have an extra comma on line 22 `process.argv[5],`. Removing it got rid of the error. However, after that I receive a `SyntaxError: Unexpected token function`, which points to line 26 `async function getRouterInfo() {`

Hi @caffeinatedMike The comma is supposed to be there. If you remove it you get the other error :) What error were you getting before? Could you please try again (and install the new package that gives more info and captures the http response. > npm i netgear)

**Comment by starkillerOG** [comment]
@MatMaul, I also have problems with pynetgear not beeing able to login to my router.
Everything is working in version 0.4.1, but after I upgraded homeassistant and got version 0.5.0 of pynetgear it failed to login to my R7000 router.

Has the fix for this issue already been merged in release 0.5.1?
I see it in the branch "new-login", but I don't see it in the master branch...
